app/src/main/java/io/appium/settings/LocationTracker.java

```diff
@@ -91,22 +108,39 @@ public void onConnectionSuspended(int i) {
 
     @Override
     public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
+        Log.e(TAG, String.format("Google Play Services location provider has failed to connect (code %s)",
+                connectionResult.toString()));
+        stopLocationUpdatesWithPlayServices();
+        googleApiStartupLatch.countDown();
     }
 
-    void start(Context mContext) {
-        if (PlayServicesHelpers.isAvailable(mContext)) {
-            Log.i(TAG, "Configuring location provider for Google Play Services");
+    synchronized void start(Context context, long googleApiConnectTimeout, boolean force) {
+        if (isStarted && !force) {
+            return;
+        }
+        isStarted = true;
+
+        if (PlayServicesHelpers.isAvailable(context)) {
+            Log.d(TAG, "Configuring location provider for Google Play Services");
             stopLocationUpdatesWithPlayServices();
-            mGoogleApiClient = new GoogleApiClient.Builder(mContext)
+            mGoogleApiClient = new GoogleApiClient.Builder(context)
                     .addApi(LocationServices.API)
                     .addConnectionCallbacks(this)
                     .addOnConnectionFailedListener(this)
                     .build();
+            googleApiStartupLatch = new CountDownLatch(1);
             mGoogleApiClient.connect();
+            if (googleApiConnectTimeout > 0) {
+                try {
+                    googleApiStartupLatch.await(googleApiConnectTimeout, TimeUnit.MILLISECONDS);
```

Comment:
If this returns `false`, `mLocationManager` should be initialized

Reply:
And since `onConnectionFailed` also calls `googleApiStartupLatch.countDown();` we need to check again if `mGoogleApiClient` is connected

Reply:
fixed